### ISSUE TITLE
Support gRPC-gateway annotations

### ISSUE THREAD

**Issue by jgall** [grpc/grpc-rust#332]
### Motivation

A frequent use-case on top of gRPC is to expose a REST API built to mirror the RPC defined in your proto file. The [grpc-gateway](https://grpc-ecosystem.github.io/grpc-gateway/) plugin for protoc does just that, automatically. I think it would be great if we could optionally support these same annotations (perhaps behind a feature flag or in a different crate?). 

### Proposal
Functional drawbacks:
- Might be considered feature-creep/bloat since it's not a part of the official proto spec. 
- This is potentially a lot of work. (Almost DEFINITELY a lot of work)

### Alternatives

One could conceivably write an entirely different crate that wraps Tonic and pre-processes the grpc-gateway annotations to create a REST API and some stubs that the user could hook in to their Tonic trait impls.

**Comment by LucioFranco** [comment]
Two main things that are somewhat blockers for this:

- Getting support for this within `prost` and exposed generically within tonic.
- Finding a way to get this support for flatbuffers as well.

I think trying to get the prost support would be a good first step!

**Comment by sleipnir** [comment]
I suggest that google grpc http transcoding annotations be used here to maintain a standard between frameworks and tools https://cloud.google.com/endpoints/docs/grpc/transcoding

**Comment by vdods** [comment]
Is there any update on this?  It would be a fantastic feature to add.

I've been trying to figure out how to do this (as a consumer of tonic, not by cracking it open), and have come across a few articles that make it seem like it should work, but I haven't had much luck.  I've been able to compile it with annotations, but issuing HTTP requests to the alleged REST endpoints oddly returns status 200 OK, but with no content, and the GRPC endpoint handler isn't being called.

Links:
- https://cloud.google.com/endpoints/docs/grpc/transcoding (formal spec; sleipnir linked this in the previous post)
- https://github.com/DazWilkin/rust-googleapis
- https://grpc.io/blog/coreos/ (talks about the transcoding abstractly but nothing concrete)
- https://earvinkayonga.com/posts/grpc-annotations-and-rust/ (this one is only about generating Swagger APIs from annotations, but I figured I'd include it as peripherally related).

**Comment by LucioFranco** [comment]
I believe the first step is to add support for this in prost. I want to say https://github.com/tokio-rs/prost/issues/301 this is the closet issue to what you may want.

**Comment by Cliftonz** [comment]
I assume that no more progress has been made on this, correct?

**Comment by LucioFranco** [comment]
Correct, we have not made much progress on the `prost` side of things.

**Comment by micolous** [comment]
I've looked this problem space in Python for a project, and am having a look at it in Rust as well (mainly because I'd rather use Rust). *This isn't suggesting that I'm planning to work on this – this is me noting down some things I found when investigating this for Python.*

You don't need to support `grpc-gateway` annotations **at all** to get gRPC-JSON transcoding functioning. A server only needs to:

* expose gRPC methods using [Google API's annotations](https://cloud.google.com/service-management/reference/rpc/google.api#http) 
* serialise/deserialise requests and responses using [ProtoJSON](https://protobuf.dev/programming-guides/json/) (rather than using `serde_json` derive macros)

That's all [Envoy gRPC-JSON transcoder](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/grpc_json_transcoder_filter) does and needs – they don't touch OpenAPI at all.

Relatedly, https://github.com/hyperium/tonic/issues/2020 is asking about allowing gRPC-Web but with ProtoJSON outputs.

---

Making OpenAPI schemas of gRPC-JSON services is a _separate_ problem, and the existing solutions have limitations and issues:

* `grpc-gateway` supports OpenAPI v2 only. There's been multiple attempts at adding OpenAPI v3 support, but none has landed at the time of writing.

  One problem with their OpenAPI v2 descriptors is that they are expressed entirely with _implicit_ presence, so that you can't _explicitly_ set any field to its default value (eg: if you set `minimum: 0` for an integer field to make it unsigned), as [this will be dropped from the final OpenAPI JSON/YAML](https://github.com/grpc-ecosystem/grpc-gateway/issues/991).

  I've observed several other things I'd call "weird", where the priority of different annotations is not well defined, and undocumented magic comment behaviour (eg: working differently when the first line of a comment ends in a `.`).

* [Google's `gnostic`](https://github.com/google/gnostic) supports OpenAPI v2 and v3. It's based on parsing the Markdown OpenAPI specifications to build a JSON schema (as there is/was no official schema), and from there build a Protobuf representation of it.

  It has the same [implicit presence bugs](https://github.com/google/gnostic/blob/ad271d568b713ad381ad6751cd8b950eade78d98/openapiv3/OpenAPIv3.proto#L552-L553) as `grpc-gateway`. It does some things a bit differently to `grpc-gateway`, but there are other bugs (null pointer dereferences, yay!), and it targets an old version of Golang (which requires significant patching to get working on current toolchains).

  While it's widely adopted (eg: GCP API gateway, Kubernetes), [the maintainer is no longer at Google](https://github.com/google/gnostic/issues/441), and it's not clear what the future is for it.

* [`solo-io/protoc-get-openapi`](https://github.com/solo-io/protoc-gen-openapi/) doesn't use protobuf `options`, but instead uses magic comments.

  I don't _think_ this has implicit field presence problems (though Golang's `encoding/json` behaves similarly to ProtoJSON), because it uses Kin's OpenAPI v3 structures, [which makes explicit-presence fields from pointers](https://github.com/getkin/kin-openapi/blob/2de45f70d4afe6446074321af3f6d542e91f7f0d/openapi3/schema.go#L113-L116).

  But this does mean there's no path to making these properly-typed Protobuf annotations.

* [`protoc-gen-jsonschema`](https://github.com/chrusty/protoc-gen-jsonschema/) just looks at the JSON Schema part (which OpenAPI's request/response schemas derive from).

  That's no longer maintained, only implements a subset of features, and has implicit presence problems.

There are a bunch of other projects around that I haven't mentioned, but they're often cribbing or forking from each other, and have the same bugs.

There's a space for someone to come up with a _good_ `protoc-gen-openapi*`, but I don't think that needs to be part of `tonic`.

The first step will be to build some _accurate_ Protobuf descriptors for OpenAPI - I wouldn't suggest copying anyone else's descriptors!

**Comment by nathanielford** [comment]
With the wind-down of Tonic development in favor of the official `grpc-rust` library, it's unlikely this will ever be directly supported by Tonic given the scope. `grpc-rust` is also not going to directly support this, as REST/JSON transcoding is outside the domain described in the specification. If you are interested in functionality like this, the recommendation would be to create a standalone crate or utilize something like Envoy proxy's `grpc_json_transcoder`.

I'm closing this issue, but let me know if you have any outstanding concerns!
